server/routes/experiments/biomed_nl/utils.py

```diff
@@ -325,3 +327,37 @@ def batch_requested_nodes(nodes, batch_size=MAX_NUM_DCID_PER_V2NODE_REQUEST):
   Splits a list of dcids into batches that do not exceed the DC API request size limit.
   """
   return [nodes[i:i + batch_size] for i in range(0, len(nodes), batch_size)]
+
+
+def cos_sim(embeddings1, embeddings2):
+  """
+    Calculates the cosine similarity between two sets of embeddings.
+
+    Args:
+        embeddings1: A numpy array of shape (n, d) where n is the number of embeddings
+                     in the first set and d is the dimensionality.
+        embeddings2: A numpy array of shape (m, d) where m is the number of embeddings
+                     in the second set and d is the dimensionality.
+
+    Returns:
+        A numpy array of shape (n, m) containing the cosine similarity scores.
+    """
+
+  embeddings1 = np.array(embeddings1)
+  embeddings2 = np.array(embeddings2)
+
+  if len(embeddings1.shape) == 1:
+    embeddings1 = embeddings1.reshape(1, -1)
+  if len(embeddings2.shape) == 1:
+    embeddings2 = embeddings2.reshape(1, -1)
+
+  # Normalize the embeddings
+  embeddings1_norm = embeddings1 / np.linalg.norm(
+      embeddings1, axis=1, keepdims=True)
+  embeddings2_norm = embeddings2 / np.linalg.norm(
+      embeddings2, axis=1, keepdims=True)
+
+  # Calculate the cosine similarity
+  similarity_matrix = np.dot(embeddings1_norm, embeddings2_norm.T)
```

Comment:
what is the `.T` at the end of embeddings2_norm for?

Reply:
transpose -- here's the original sentence_transformers cos_sim code ([link](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/util.py#L92-L108)) ; they also take the transpose

it's so that the final similarity scores are in the right order. since we only use this for 1 embedding vs 1 embedding, it doesn't really make a difference because the final output is just a 1x1 "matrix"

Reply:
would it make more sense to change it to a single vector vs vector comparison? 

i kept the matrix thing just to be consistent with sentence_transformers, but we just need single scores in the current implementation

Reply:
oh interesting, I was just confused because the `.T` was only added to one of the embeddings, but if that's how sentence_transformers does it, let's just keep that! Let's be consistent with sentence_transformers for now

Reply:
sg! thanks for the review!!